py/server.py

```diff
@@ -695,6 +696,94 @@ def gather_envs(state):
     return sorted(list(set(items + list(state.keys()))))
 
 
+def compare_envs(state, eids, socket):
+    logging.info('comparing envs')
+    eidNums = {e:str(i) for i,e in enumerate(eids)}
+    env = {}
+    envs = {}
+    for eid in eids:
+        if eid in state:
+            envs[eid] = state.get(eid)
+        else:
+            p = os.path.join(FLAGS.env_path, eid.strip(), '.json')
+            if os.path.exists(p):
+                env = tornado.escape.json_decode(open(p, 'r').read())
+                state[eid] = env
+                envs[eid] = env
+
+
+
+    res = copy.deepcopy(envs[list(envs.keys())[0]])
+    name2Wid = {res['jsons'][wid].get('title', None):wid+'_compare' for wid in res.get('jsons', {})
+                if 'title' in res['jsons'][wid]}
+    for wid in list(res['jsons'].keys()):
+        res['jsons'][wid+'_compare'] = res['jsons'][wid]
+        res['jsons'][wid] = None
+        res['jsons'].pop(wid)
+
+    for ix, eid in enumerate(envs.keys()):
+        env = envs[eid]
+        for wid in env.get('jsons',{}).keys():
+            win = env['jsons'][wid]
+            if win.get('type', None) != 'plot': continue
+            if 'content' not in win: continue
+            if 'title' not in win: continue
+            title = win['title']
+            if title not in name2Wid: continue
+
+            destWid = name2Wid[title]
+
+            if ix == 0:
+                res['jsons'][destWid]['has_compare'] = False
+                res['jsons'][destWid]['content']['layout']['showlegend'] = True
+                for di,d in enumerate(res['jsons'][destWid]['content']['data']):
+                    res['jsons'][destWid]['content']['data'][di]['name'] = '{}_{}'.format(eidNums[eid], d['name'])
+            else:
+                res['jsons'][destWid]['has_compare'] = True
+                for di,d in enumerate(win['content']['data']):
+                    d = copy.deepcopy(d)
+                    d['name'] = '{}_{}'.format(eidNums[eid], d['name'])
+                    res['jsons'][destWid]['content']['data'].append(d)
+
+    for k in res['jsons']:
```

Comment:
If I understand this properly, the compare construction will only display windows that are shared in title by at least 2 of the environments being compared? Can you add comments here and above to display your logic? 

(It's a nice design decision though 👍 )